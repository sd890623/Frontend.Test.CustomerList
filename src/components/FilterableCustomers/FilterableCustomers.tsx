import {FC, useCallback, useState} from "react";
import {CustomerState, ICustomer} from "../../types/types";
import {shallowEqual, useSelector} from "react-redux";
import {Customer} from "../Customer/Customer";
import {removeCustomer} from "../../redux/actions/customerActions";
import {
    StyledFilterableCustomers,
    StyledFilteredCustomers,
    StyledInput,
    StyledLabel
} from "./StyledFilterableCustomers";

export const getFilteredCustomers = (word:string, customers:ICustomer[]):ICustomer[]=>{
    const toUpperCaseWord=word.toUpperCase()
    return customers.filter(customer=>{
        return customer.firstName.toUpperCase().includes(toUpperCaseWord)
            || customer.lastName.toUpperCase().includes(toUpperCaseWord)
            || customer.phoneNumber.toUpperCase().includes(toUpperCaseWord)
    })
}
const FilterableCustomers: FC = ()=>{
    const [searchWord,setSearchWord]=useState('')
    const customers: ICustomer[] = useSelector(
        (state: CustomerState) => state.customers,
        shallowEqual
    );
    const getFilteredCustomersCallback = useCallback(()=>getFilteredCustomers(searchWord,customers),[searchWord,customers])

    return <StyledFilterableCustomers>
        <StyledLabel>
            Name or mobile
        </StyledLabel>
        <StyledInput
            data-testid="input"
            name="search-word"
            type="text"
            value={searchWord}
            onChange={e=>setSearchWord(e.target.value)}
        />
        <StyledFilteredCustomers>{getFilteredCustomersCallback().map((customer: ICustomer) => (
            <Customer
                key={customer.id}
                customer={customer}
                removeCustomer={removeCustomer}
            />
        ))}
        </StyledFilteredCustomers>
    </StyledFilterableCustomers>
}
export default FilterableCustomers