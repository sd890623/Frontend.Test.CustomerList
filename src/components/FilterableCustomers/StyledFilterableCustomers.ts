import styled from "styled-components";

export const StyledFilterableCustomers=styled.div`
  width:100%;
  margin-top:1rem;
`
export const StyledLabel = styled.label`
  padding-left: 1rem;
  margin-right: 1rem;
`;

export const StyledInput = styled.input``;

export const StyledFilteredCustomers = styled.div``;