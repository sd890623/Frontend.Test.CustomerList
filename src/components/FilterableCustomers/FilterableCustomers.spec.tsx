import {fireEvent} from '@testing-library/react';
import { render } from "../../utils/testUtils";
import FilterableCustomers, {getFilteredCustomers} from "./FilterableCustomers";
import {mockCustomers} from "../../mocks/tests/Components/Customers";

describe("<FilterableCustomers />", () => {
    it("should render a <FilterableCustomers />", () => {
        const wrapper = render(<FilterableCustomers  />);
        expect(wrapper.container).toMatchSnapshot();
    });
    it("should have getFilteredCustomers filter results", () => {
        expect(getFilteredCustomers('alan',mockCustomers)).toEqual([{
            id: 2,
            firstName: "Alan",
            lastName: "Turing",
            phoneNumber: "(03) 9599 1234",
        }])
        expect(getFilteredCustomers('love',mockCustomers)).toEqual([{
            id: 3,
            firstName: "Ada",
            lastName: "Lovelace",
            phoneNumber: "+61 423 345 567",
        }])
        expect(getFilteredCustomers('',mockCustomers).length).toBe(3)
        expect(getFilteredCustomers('abcd',mockCustomers).length).toBe(0)

    });
    it("should type search words in input", () => {
        const {getByTestId} = render(<FilterableCustomers  />);
        const input =getByTestId('input') as HTMLInputElement
        expect(input.value).toBe('')
        fireEvent.change(input, {target: {value: 'alan'}})
        expect(input.value).toBe('alan')
        fireEvent.change(input, {target: {value: ''}})
        expect(input.value).toBe('')
    });
    it("should type search words and see customers list change", () => {
        const {getByTestId,getAllByTestId} = render(<FilterableCustomers  />);
        const input =getByTestId('input')
        expect(getAllByTestId('customer').length).toBe(3)
        fireEvent.change(input, {target: {value: 'alan'}})
        expect(getAllByTestId('customer').length).toBe(1)
        expect(getByTestId('customer').firstChild?.textContent).toBe('Alan Turing')
        fireEvent.change(input, {target: {value: ''}})
        expect(getAllByTestId('customer').length).toBe(3)
    });
});
