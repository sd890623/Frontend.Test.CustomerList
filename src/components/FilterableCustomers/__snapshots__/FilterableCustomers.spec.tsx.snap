// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FilterableCustomers /> should render a <FilterableCustomers /> 1`] = `
<div>
  <div
    class="sc-eCApnc fgVDfB"
  >
    <label
      class="sc-jSFjdj joFUgd"
    >
      Name or mobile
    </label>
    <input
      class="sc-gKAaRy"
      data-testid="input"
      name="search-word"
      type="text"
      value=""
    />
    <div
      class="sc-iCoGMd"
    >
      <div
        class="sc-bdnxRM llOSvQ"
        data-testid="customer"
      >
        <h2
          class="sc-gtsrHT bmgxMS"
        >
          Charles
           
          Babbage
        </h2>
        <p
          class="sc-dlnjwi kUbGey"
        >
          Phone number: 
          0412 123 123
        </p>
        <button
          class="sc-hKFxyN dwHgfV"
        >
          Delete
        </button>
      </div>
      <div
        class="sc-bdnxRM llOSvQ"
        data-testid="customer"
      >
        <h2
          class="sc-gtsrHT bmgxMS"
        >
          Alan
           
          Turing
        </h2>
        <p
          class="sc-dlnjwi kUbGey"
        >
          Phone number: 
          (03) 9599 1234
        </p>
        <button
          class="sc-hKFxyN dwHgfV"
        >
          Delete
        </button>
      </div>
      <div
        class="sc-bdnxRM llOSvQ"
        data-testid="customer"
      >
        <h2
          class="sc-gtsrHT bmgxMS"
        >
          Ada
           
          Lovelace
        </h2>
        <p
          class="sc-dlnjwi kUbGey"
        >
          Phone number: 
          +61 423 345 567
        </p>
        <button
          class="sc-hKFxyN dwHgfV"
        >
          Delete
        </button>
      </div>
    </div>
  </div>
</div>
`;
